node/src/Unison/Runtime/ResourcePool.hs

```diff
@@ -0,0 +1,64 @@
+module Unison.Runtime.ResourcePool where
+
+import qualified Data.Map as M
+import qualified Control.Concurrent.MVar as MVar
+import Data.Time (UTCTime, getCurrentTime, addUTCTime)
+import qualified Control.Concurrent as CC
+
+-- acquire returns the resource, and the cleanup action ("finalizer") for that resource
+data Pool p r = Pool { acquire :: p -> Int -> IO (r, IO ()) }
+
+type ResourceKey p = (p,CC.ThreadId)
+type Cache p r = MVar.MVar (M.Map (ResourceKey p) (r, UTCTime, IO ()))
+
+addResourceToMap :: (Ord p) =>  (r -> IO()) -> Cache p r -> p -> r -> Int -> Int -> IO(CC.ThreadId) -> IO ()
+addResourceToMap releaser mVarCache p r wait maxPoolSize getThread = do
+  now <- getCurrentTime
+  threadId <- getThread
+  let expiry = addUTCTime (fromIntegral wait) now
+  cacheMap <- MVar.takeMVar mVarCache
```

Comment:
Use `modifyMVar_` - lets you safely modify an `MVar`, putting back the original value if an exception occurs. The code you have here is missing exception handling, so if `release r` bombs the `MVar` is left empty.